src/core/Note.cpp

```diff
@@ -232,3 +232,32 @@ bool Note::withinRange(int tickStart, int tickEnd) const
 	return pos().getTicks() >= tickStart && pos().getTicks() <= tickEnd
 		&& length().getTicks() != 0;
 }
+
+
+
+
+/*! \brief Get the start/end/bottom/top positions of notes in a vector
+ *
+ *  Returns no value if there are no notes
+ */
+std::optional<NoteBounds> boundsForNotes(const NoteVector& notes)
+{
+	if (notes.empty()) { return {}; }
+
+	TimePos start = notes.front()->pos();
+	TimePos end = start;
+	int lower = notes.front()->key();
+	int upper = lower;
+
+	for (const Note* note: notes)
+	{
+		// TODO should we assume that NoteVector is always sorted correctly,
+		// so first() always has the lowest time position?
+		start = std::min(start, note->pos());
+		end = std::max(end, note->endPos());
+		lower = std::min(lower, note->key());
+		upper = std::max(upper, note->key());
+	}
+
+	return NoteBounds{start, end, lower, upper};
```

Comment:
Just for my understanding: A negative side effect of returning `std::optional<NoteBounds>` instead of passing the four parameters via references (as "out parameters") is, using the return value might cause a copy? However, in C++17, it seems to be [forced to elide the copying](https://en.cppreference.com/w/cpp/language/copy_elision), which is why we do not need to worry?

Reply:
Interesting. Even prior to C++17
> compilers are permitted, but not required to omit the copy

I did some tests. With `optional` it's a different story, because the value seems to be constructed first and then copied into the `optional` object. To avoid it you must do

    return std::optional<NoteBounds>{std::in_place, start, end, lower, upper};

I'm feeling tempted to remove all the `optional` fuzz and just go with `operator bool()`

Reply:
At -O1 or above, GCC optimises out the extra copy and stores the constructor arguments straight into the optional. (See https://godbolt.org/z/aKoWcrGnY.) Even if it didn't, it's only copying 16 bytes in a hot memory region. An optional is better semantically than giving a special meaning to a particular value of `NoteBounds`, and I wouldn't sacrifice that unless profiling showed it to be especially problematic.

Reply:
Agreed that this is not a problem, disagreeing on semantics. But that's why you're the C++ guy and I'm not (cool site btw)